src/RoomsSDKAdapter.js

```diff
@@ -36,42 +37,67 @@ export default class RoomsSDKAdapter extends RoomsAdapter {
     };
   }
 
+  /**
+   * Create an observable that subscribes to room change events emitted via websocket and
+   * emits the changed room object.
+   *
+   * @private
+   * @param {string} ID ID of room to get
+   * @returns {Observable.<Room>} Observable that emits changes to the room
+   * @memberof RoomsSDKAdapter
+   */
+  addRoomChangeListener(ID) {
+    const roomChange$ = fromEvent(this.datasource.rooms, ROOM_UPDATED_EVENT);
+
+    return roomChange$.pipe(
+      // Is the room change event for our subscribed room?
+      filter((event) => event.data.id === ID),
+      // Event data doesn't have the room data in it, so we need to fetch manually
+      mergeMap(async () => this.fetchRoom(ID))
+    );
+  }
+
+  /**
+   * Builds the observable flow for fetching room data, then listening for
+   * websocket events about room changes.
+   *
+   * @private
+   * @param {string} ID
+   * @returns {Observable.<Room>}
+   * @memberof RoomsSDKAdapter
+   */
+  getRoomObservable(ID) {
+    const room$ = from(this.fetchRoom(ID));
+
+    return concat(
+      // Fetch Our Room Data
+      room$,
+      this.addRoomChangeListener(ID)
+    );
+  }
+
   /**
    * Returns an observable that emits room data of the given ID.
    *
+   * @public
    * @param {string} ID ID of room to get
    * @returns {Observable.<Room>}
    * @memberof RoomsSDKAdapter
    */
   getRoom(ID) {
     if (!(ID in this.getRoomObservables)) {
-      const source = Observable.create((observer) => {
-        // Start listening for room changes
-        this.datasource.rooms.listen();
-        this.datasource.rooms.on(ROOM_UPDATED_EVENT, () => {
-          // Room has updates, fetch and send
-          this.fetchRoom(ID)
-            .then((room) => observer.next(room))
-            .catch((error) => observer.error(error));
-        });
+      // Tell the sdk to start listening to room changes
+      this.datasource.rooms.listen();
 
-        // Get our initial room from the SDK
-        this.fetchRoom(ID)
-          .then((room) => observer.next(room))
-          .catch((error) => observer.error(error));
-
-        return () => {
-          // Cleanup when subscription count is 0
-          this.datasource.rooms.stopListening();
-          this.datasource.rooms.off(ROOM_UPDATED_EVENT);
-          delete this.getRoomObservables[ID];
-        };
-      });
+      const getRoom$ = this.getRoomObservable(ID);
```

Comment:
I appreciate trying to make the code more modular, but let's be careful of over-engineering. I think the logic for this observable is so complex that needs to be broken down in 3 methods. See my [sample attempt](https://github.com/lalli-flores/sdk-component-adapter/blob/rooms-remove-async/src/RoomsSDKAdapter.js#L63)

Reply:
There's so much logic in the `getRoomObservable` function, it is easier to comprehend by splitting it into its own function.

Reply:
I see the that you take the needed extra work to a different method, where I just left mine in the same method. We just need to come up with a conclusion on how we want to implement this. I personally don't care *much*, though I'm leaning a bit towards my approach to make the code looks simpler and easier to read and have all the observables together to look for.

Reply:
but whatever pattern we are coming up with, needs to be applied in both PRs.